File: apps/mobile/src/store/meta.ts
/**
 * Vocabularies from /meta. Loaded once per app session; safe fallbacks keep the
 * pickers usable offline.
 */
import { create } from 'zustand';
import { useI18nStore } from '../i18n';
import { metaApi } from '../services';
import { cache } from '../services/cache';
import type { City, WardrobeOptions } from '../services/types';

const FALLBACK_OPTIONS: WardrobeOptions = {
  garment_types: [],
  fabrics: [],
  occasions: [],
  seasons: [
    { slug: 'summer', label: 'Summer' },
    { slug: 'monsoon', label: 'Monsoon' },
    { slug: 'winter', label: 'Winter' },
    { slug: 'all_season', label: 'All season' },
  ],
  colors: [],
  regional_styles: [
    { slug: 'rajasthani', label: 'Rajasthani' },
    { slug: 'south_indian', label: 'South Indian' },
    { slug: 'punjabi', label: 'Punjabi' },
    { slug: 'mumbai_minimal', label: 'Mumbai Minimal' },
    { slug: 'pan_india_fusion', label: 'Pan-India Fusion' },
  ],
  conditions: [
    { slug: 'new', label: 'New' },
    { slug: 'good', label: 'Good' },
    { slug: 'worn', label: 'Worn' },
  ],
};

interface MetaState {
  options: WardrobeOptions;
  cities: City[];
  loaded: boolean;
  loadedLang: string | null;
  load: (force?: boolean) => Promise<void>;
}

export const useMetaStore = create<MetaState>((set, get) => ({
  options: FALLBACK_OPTIONS,
  cities: [],
  loaded: false,
  loadedLang: null,
  /** Loads vocab for the current UI language: cache first (instant, offline), then network. */
  load: async (force = false) => {
    const lang = useI18nStore.getState().language;
    if (get().loaded && get().loadedLang === lang && !force) return;
    const key = `meta.${lang}`;
    const hit = await cache.get<{ options: WardrobeOptions; cities: City[] }>(key);
    if (hit) set({ options: hit.data.options, cities: hit.data.cities, loaded: true, loadedLang: lang });
    try {
      const [options, cities] = await Promise.all([metaApi.wardrobeOptions(lang), metaApi.cities()]);
      set({ options, cities, loaded: true, loadedLang: lang });
      void cache.set(key, { options, cities });
    } catch {
      // keep cache/fallbacks; retry next call
    }
  },
}));

// re-fetch vocab whenever the language changes
useI18nStore.subscribe((s, prev) => {
  if (s.language !== prev.language) void useMetaStore.getState().load(true);
});

export const labelFor = (opts: { slug: string; label: string }[], slug: string | null | undefined) =>
  opts.find((o) => o.slug === slug)?.label ?? (slug ? slug.replace(/_/g, ' ') : '—');

/** Garment-type chips for a user: their gender's pieces + unisex; 'other' sees everything. */
export const garmentTypesFor = (opts: { slug: string; label: string; gender?: string }[], gender: string | undefined) => {
  const allowed = gender === 'female' ? ['women', 'unisex'] : gender === 'male' ? ['men', 'unisex'] : null;
  return allowed ? opts.filter((o) => !o.gender || allowed.includes(o.gender)) : opts;
};

File: apps/api/app/routers/meta.py
"""
/meta — vocabularies from the knowledge base, so clients never hard-code them.
"""
from fastapi import APIRouter

from app import knowledge
from app.models.garment import GarmentCondition
from app.schemas.meta import BodyTypeOut, CityOut, GarmentTypeOption, Option, WardrobeOptionsOut

router = APIRouter(prefix="/meta", tags=["meta"])

_STYLE_LABELS = {
    "rajasthani": "Rajasthani",
    "south_indian": "South Indian",
    "punjabi": "Punjabi",
    "mumbai_minimal": "Mumbai Minimal",
    "pan_india_fusion": "Pan-India Fusion",
}
_SEASON_LABELS = {
    "summer": "Summer",
    "monsoon": "Monsoon",
    "winter": "Winter",
    "all_season": "All season",
}


def _title(slug: str) -> str:
    return slug.replace("_", " ").title()


@router.get("/wardrobe-options", response_model=WardrobeOptionsOut)
async def wardrobe_options() -> WardrobeOptionsOut:
    return WardrobeOptionsOut(
        garment_types=[
            GarmentTypeOption(slug=g["slug"], label=g["label"], gender=g.get("gender", "unisex"))
            for g in knowledge.garment_types()
        ]
        + [GarmentTypeOption(slug="other", label="Other", gender="unisex")],
        fabrics=[Option(slug=s, label=_title(s)) for s in knowledge.fabric_slugs()]
        + [Option(slug="other", label="Other")],
        occasions=[Option(slug=o["slug"], label=o["label"]) for o in knowledge.occasions()],
        seasons=[Option(slug=s, label=_SEASON_LABELS[s]) for s in knowledge.SEASONS],
        colors=[Option(slug=c, label=_title(c)) for c in knowledge.COLORS],
        regional_styles=[Option(slug=s, label=_STYLE_LABELS[s]) for s in knowledge.REGIONAL_STYLES],
        conditions=[Option(slug=c.value, label=_title(c.value)) for c in GarmentCondition],
    )


@router.get("/cities", response_model=list[CityOut])
async def cities() -> list[CityOut]:
    return [CityOut(**c) for c in knowledge.cities()]


@router.get("/body-types", response_model=list[BodyTypeOut])
async def body_types(gender: str | None = None) -> list[BodyTypeOut]:
    """Body types (with fit guidance) for onboarding; men and women differ."""
    return [BodyTypeOut(**b) for b in knowledge.body_types_for(gender)]
